File: sdks/go/dcp/v2/crypto_provider.go
package v2

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
)

// CryptoProvider defines the interface for signature-based cryptographic operations.
type CryptoProvider interface {
	Alg() string
	KeySize() int
	SigSize() int
	IsConstantTime() bool
	GenerateKeypair() (*GeneratedKeypair, error)
	Sign(message []byte, secretKeyB64 string) ([]byte, error)
	Verify(message []byte, signature []byte, publicKeyB64 string) (bool, error)
}

// GeneratedKeypair holds a freshly generated keypair with its derived kid.
type GeneratedKeypair struct {
	Kid          string
	PublicKeyB64 string
	SecretKeyB64 string
}

// KemProvider defines the interface for Key Encapsulation Mechanism operations.
type KemProvider interface {
	Alg() string
	GenerateKeypair() (*KemKeypair, error)
	Encapsulate(publicKeyB64 string) (*EncapsulateResult, error)
	Decapsulate(ciphertextB64 string, secretKeyB64 string) ([]byte, error)
}

// KemKeypair holds a KEM keypair.
type KemKeypair struct {
	Kid          string
	PublicKeyB64 string
	SecretKeyB64 string
}

// EncapsulateResult holds the output of a KEM encapsulation.
type EncapsulateResult struct {
	CiphertextB64  string
	SharedSecretB64 string
}

// KeyEntry represents a public key entry in a V2 artifact.
type KeyEntry struct {
	Kid          string  `json:"kid"`
	Alg          string  `json:"alg"`
	PublicKeyB64 string  `json:"public_key_b64"`
	CreatedAt    string  `json:"created_at"`
	ExpiresAt    *string `json:"expires_at"`
	Status       string  `json:"status"`
}

// DeriveKid computes a deterministic key identifier:
// kid = hex(SHA-256(UTF8(alg) || 0x00 || publicKeyBytes))[0:32]
func DeriveKid(alg string, publicKeyBytes []byte) string {
	h := sha256.New()
	h.Write([]byte(alg))
	h.Write([]byte{0x00})
	h.Write(publicKeyBytes)
	digest := h.Sum(nil)
	full := hex.EncodeToString(digest)
	if len(full) < 32 {
		return full
	}
	return full[:32]
}

// ErrUnknownAlgorithm is returned when a requested algorithm is not registered.
type ErrUnknownAlgorithm struct {
	Alg  string
	Kind string
}

func (e *ErrUnknownAlgorithm) Error() string {
	return fmt.Sprintf("unknown %s algorithm: %s", e.Kind, e.Alg)
}

File: sdks/typescript/src/core/crypto-provider.ts
/**
 * DCP v2.0 CryptoProvider and KemProvider interfaces.
 *
 * All signature algorithms and KEMs are accessed through these interfaces,
 * enabling crypto-agility: swap Ed25519 for ML-DSA-65 without changing
 * calling code.
 */

import { createHash } from 'crypto';
import type { CompositeSignature } from './composite-sig.js';

// ── Signature provider ──

export interface CryptoProvider {
  readonly alg: string;
  readonly keySize: number;
  readonly sigSize: number;
  readonly isConstantTime: boolean;

  generateKeypair(): Promise<{
    kid: string;
    publicKeyB64: string;
    secretKeyB64: string;
  }>;
  sign(message: Uint8Array, secretKeyB64: string): Promise<Uint8Array>;
  verify(
    message: Uint8Array,
    signature: Uint8Array,
    publicKeyB64: string,
  ): Promise<boolean>;
}

// ── KEM provider (optional, for encryption) ──

export interface KemProvider {
  readonly alg: string;

  generateKeypair(): Promise<{
    publicKeyB64: string;
    secretKeyB64: string;
  }>;
  encapsulate(publicKeyB64: string): Promise<{
    sharedSecret: Uint8Array;
    ciphertextB64: string;
  }>;
  decapsulate(
    ciphertextB64: string,
    secretKeyB64: string,
  ): Promise<Uint8Array>;
}

// ── Composite operations ──

export interface CompositeOps {
  compositeSign(
    context: string,
    payload: Uint8Array,
    classicalSecret: string,
    pqSecret: string,
  ): Promise<CompositeSignature>;

  compositeVerify(
    context: string,
    payload: Uint8Array,
    compositeSig: CompositeSignature,
    classicalPubkey: string,
    pqPubkey: string,
  ): Promise<{
    valid: boolean;
    classical_valid: boolean;
    pq_valid: boolean;
  }>;
}

// ── Key entry (V2 key object) ──

export type KeyStatus = 'active' | 'revoked' | 'expired';

export interface KeyEntry {
  kid: string;
  alg: string;
  public_key_b64: string;
  created_at: string;
  expires_at: string | null;
  status: KeyStatus;
}

// ── Deterministic kid derivation ──

/**
 * Derive a deterministic key identifier from algorithm name and public key bytes.
 *
 * kid = hex(SHA-256(UTF8(alg) || 0x00 || raw_public_key_bytes))[0:32]
 *
 * This guarantees uniqueness and determinism: any SDK can recompute the kid.
 */
export function deriveKid(alg: string, publicKeyBytes: Uint8Array): string {
  const encoder = new TextEncoder();
  const algBytes = encoder.encode(alg);
  const separator = new Uint8Array([0x00]);
  const input = new Uint8Array(
    algBytes.length + 1 + publicKeyBytes.length,
  );
  input.set(algBytes, 0);
  input.set(separator, algBytes.length);
  input.set(publicKeyBytes, algBytes.length + 1);
  return createHash('sha256').update(input).digest('hex').slice(0, 32);
}
